File: lectures/210922_performance_and_memory/code/collatz.c
#include <stdio.h>
#include <stdlib.h>
#include <assert.h>

int collatz(int n) {
  int i = 0;
  while (n != 1) {
    if (n % 2 == 0) { n = n / 2; }
    else            { n = 3 * n + 1; }
    i++;
  }
  return i;
}

int main(int argc, char** argv) {
  assert(argc == 2);
  int k = atoi(argv[1]);
  for (int n = 1; n < k; n++) {
    printf("%d %d\n", n, collatz(n));
  }
}

File: lectures/210922_performance_and_memory/code/collatz.py
import sys

def collatz(n):
    i = 0
    while n != 1:
        if n % 2 == 0:
            n = n // 2
        else:
            n = 3 * n + 1
        i = i + 1
    return i

k = int(sys.argv[1])
for n in range(1, k):
    print(n, collatz(n))

